### PR TITLE
Refactor/downloading

### PR DESCRIPTION
**Pull request by CremaLuca**
## Premise

Downloading needed a deep refactor. Insights in [this trello card](https://trello.com/c/g9olyjV1/159-refactor-del-download)

### `otri/downloader`

Moved all the superclasses in the `__init__.py` file.

#### Requests Limiter

Defines a class that handles how many requests a provider can handle and how frequently. It is called before performing a request to check whether the downloader needs to wait sometime before checking for availability again.

Two kinds of limiters have been implemented:

- The generic `DefaultRequestsLimiter` defines how many requests every how much time (eg. 5 requests per minute or 1 request every 0.2 seconds)
- The Tradier's one uses the response data to read from the headers how many requests there are left in the current time period.

#### Unified Intervals

We needed a way to represent intervals independently from the provider, so an enum-like constants list has been implemented. Not sure it's the cleanest solution but I've been thinking about it day and night since I started refactoring.

#### Aliasing

All providers must alias their fields to match the other providers AND filter the fields to load only the useful ones. This has been done by defining an aliases dictionary where the key is the target name and the value is the provider data's name.
I couldn't come up with an idea to make sure all providers give the same name to the same fields, given that we don't know every possible field for data I don't think that's possible.

#### Date formatting

Providers could give date-times in different formats (not GMT, epoch, or hh:mm:ss DD-MM-YY). The downloader converts the datetime filed in a unified version of the datetime by calling a conversion function that can be overridden by passing it to the `set_datetime_formatter` method.

#### Max attempts

Some providers like yahoo finance may sometimes refuse to give data at the first request and might need a second one. So every downloader can now define a `max_attempts` constant that defines how many attempts has to be made before giving up the download of that single thing (can be a ticker or a group of tickers).

#### Pre/Post process

The processing phase is defined as the aliasing and the date formatting. The provider might want to add some processing before or after the said phase. It can be done by overriding two pre/post-processing methods.

#### Request timeformat

It MIGHT has added a little bit of complexity but every provider of time-series requires a start and end date as a string with different formats between providers. A passed function formats the datetime before passing it to the data request method.

## Picture of A ~cute animal~ mong us

Emergency meeting

### CODE REVIEW

File: test/downloader/download_limiter_test.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    def setUp(self):
+        self.limiter = DefaultRequestsLimiter(10, timedelta(seconds=1))
+
+    def test_on_request_resets(self):
+        self.limiter._on_request()
+        self.assertEqual(1, self.limiter.request_counter)
+        self.assertLessEqual(datetime.utcnow(), self.limiter.next_reset)
+
+    def test_wait_time_not_zero(self):
+        for i in range(10):
```
Comment (CremaLuca) [inline test/downloader/download_limiter_test.py:17]: Codacy found an issue: [Unused variable 'i'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- otri/downloader/__init__.py  14
- otri/downloader/yahoo.py  5
- otri/downloader/alpha_vantage.py  6
- option_download.py  3
- timeseries_download.py  3
- realtime_download.py  1
- test/downloader/download_limiter_test.py  2
         

Complexity decreasing per file
==============================
+ otri/downloader/tradier.py  -5
         

Clones added
============
- otri/downloader/__init__.py  4
- otri/downloader/tradier.py  7
- option_download.py  6
- timeseries_download.py  3
- realtime_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191&bid=20226750)

File: realtime_download.py [hunk 2]
```diff
@@ -43,27 +43,30 @@ def run(self):
         self.execute = True
         while(self.execute or self.contents_queue.qsize() != 0):
             # Wait at most <timeout> seconds for something to pop in the queue
-            contents = self.contents_queue.get(block=True, timeout=30)
-            log.d("attempting to upload contents")
             try:
-                self.importer.from_contents(contents)
+                contents = self.contents_queue.get(block=True, timeout=5)
+            except Exception:
```
Comment (CremaLuca) [inline realtime_download.py:48]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: timeseries_download.py [hunk 3]
```diff
@@ -17,89 +17,139 @@
...
-                md_row.data_json['provider'].append(self.provider_name)
-            log.d("updated ticker provider")
-        log.d("successfully uploaded {}".format(self.ticker))
+        log.i("started uploader worker")
+        self.execute = True
+        while(self.execute or self.contents_queue.qsize() != 0):
+            # Wait at most <timeout> seconds for something to pop in the queue
+            try:
+                contents = self.contents_queue.get(block=True, timeout=5)
+            except Exception:
```
Comment (CremaLuca) [inline timeseries_download.py:98]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: option_download.py [hunk 4]
```diff
@@ -14,109 +14,153 @@
...
+        self.importer = importer
+
+    def run(self):
+        log.i("started uploader worker")
+        self.execute = True
+        while(self.execute or self.contents_queue.qsize() != 0):
+            # Wait at most <timeout> seconds for something to pop in the queue
+            try:
+                contents = self.contents_queue.get(block=True, timeout=5)
+            except Exception:
```
Comment (CremaLuca) [inline option_download.py:120]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- timeseries_download.py  3
- otri/downloader/yahoo.py  5
- test/downloader/download_limiter_test.py  2
- otri/downloader/__init__.py  14
- realtime_download.py  1
- otri/downloader/alpha_vantage.py  6
- option_download.py  3
         

Complexity decreasing per file
==============================
+ otri/downloader/tradier.py  -5
         

Clones added
============
- timeseries_download.py  3
- otri/downloader/__init__.py  4
- otri/downloader/tradier.py  7
- realtime_download.py  3
- option_download.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191&bid=20226750)

File: realtime_download.py (hunk 2 above)
Comment (CremaLuca) [inline realtime_download.py:48]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: timeseries_download.py (hunk 3 above)
Comment (CremaLuca) [inline timeseries_download.py:98]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: option_download.py (hunk 4 above)
Comment (CremaLuca) [inline option_download.py:120]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: test/downloader/download_limiter_test.py (hunk 1 above)
Comment (CremaLuca) [inline test/downloader/download_limiter_test.py:17]: Codacy found an issue: [Unused variable 'i'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- test/downloader/download_limiter_test.py  2
- otri/downloader/alpha_vantage.py  6
- timeseries_download.py  3
- option_download.py  3
- otri/downloader/yahoo.py  5
- otri/downloader/__init__.py  15
- realtime_download.py  1
         

Complexity decreasing per file
==============================
+ otri/downloader/tradier.py  -5
         

Clones added
============
- otri/downloader/tradier.py  7
- timeseries_download.py  3
- option_download.py  6
- otri/downloader/__init__.py  4
- realtime_download.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191&bid=20226750)

File: realtime_download.py (hunk 2 above)
Comment (CremaLuca) [inline realtime_download.py:48]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: timeseries_download.py (hunk 3 above)
Comment (CremaLuca) [inline timeseries_download.py:98]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: option_download.py (hunk 4 above)
Comment (CremaLuca) [inline option_download.py:120]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: test/downloader/download_limiter_test.py (hunk 1 above)
Comment (CremaLuca) [inline test/downloader/download_limiter_test.py:17]: Codacy found an issue: [Unused variable 'i'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

**Comment by CremaLuca** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- otri/downloader/alpha_vantage.py  6
- otri/downloader/__init__.py  15
- timeseries_download.py  3
- test/downloader/download_limiter_test.py  2
- otri/downloader/yahoo.py  5
- realtime_download.py  1
- option_download.py  3
         

Complexity decreasing per file
==============================
+ otri/downloader/tradier.py  -5
         

Clones added
============
- otri/downloader/__init__.py  4
- otri/downloader/tradier.py  7
- timeseries_download.py  3
- realtime_download.py  3
- option_download.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191&bid=20226750)

File: realtime_download.py (hunk 2 above)
Comment (CremaLuca) [inline realtime_download.py:48]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: timeseries_download.py (hunk 3 above)
Comment (CremaLuca) [inline timeseries_download.py:98]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: option_download.py (hunk 4 above)
Comment (CremaLuca) [inline option_download.py:120]: Codacy found an issue: [Try, Except, Continue detected.](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

File: test/downloader/download_limiter_test.py (hunk 1 above)
Comment (CremaLuca) [inline test/downloader/download_limiter_test.py:17]: Codacy found an issue: [Unused variable 'i'](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6201191)

**Review by ricdezen — APPROVED** [review]

**Pull request merged**